File: frontend/src/types/referenceSchema.ts
/**
 * Ответ `GET /api/v1/economics/references/schema`.
 *
 * Фронт не хранит собственных знаний о полях payload: состав, единицы,
 * ссылки на другие разделы и группировка формы приходят с сервера.
 */

export type JsonSchemaNode = {
  type?: string;
  anyOf?: JsonSchemaNode[];
  enum?: string[];
  items?: JsonSchemaNode;
  $ref?: string;
  title?: string;
  description?: string;
  default?: unknown;
  minimum?: number;
  maximum?: number;
  format?: string;
  pattern?: string;
  properties?: Record<string, JsonSchemaNode>;
  additionalProperties?: boolean | JsonSchemaNode;
  "x-unit"?: string;
  "x-ref"?: string;
  "x-internal"?: boolean;
};

export type JsonSchemaObject = JsonSchemaNode & {
  properties?: Record<string, JsonSchemaNode>;
  $defs?: Record<string, JsonSchemaNode>;
  required?: string[];
};

export type ReferenceFieldset = {
  title: string;
  fields: string[];
};

export type ReferenceSectionSchema = {
  code: string;
  label: string;
  group: string;
  view: "table" | "matrix";
  deprecated: boolean;
  list_columns: string[];
  fieldsets: ReferenceFieldset[];
  json_schema: JsonSchemaObject;
};

export type ReferenceSchemaCatalog = {
  groups: Array<{ code: string; label: string }>;
  sections: Record<string, ReferenceSectionSchema>;
};

File: api/schemas/reference_schema.py
"""Ответ `GET /economics/references/schema`.

Фронт рисует формы справочников по этому ответу и не хранит собственных знаний
о полях: состав, единицы, ссылки и группировка приходят с сервера.
"""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

__all__ = ["ReferenceGroupSchema", "ReferenceSectionSchema", "ReferenceSchemaResponse"]


class ReferenceGroupSchema(BaseModel):
    code: str
    label: str


class ReferenceFieldsetSchema(BaseModel):
    title: str
    fields: list[str] = Field(default_factory=list)


class ReferenceSectionSchema(BaseModel):
    code: str
    label: str
    group: str
    view: str = "table"
    deprecated: bool = False
    list_columns: list[str] = Field(default_factory=list)
    fieldsets: list[ReferenceFieldsetSchema] = Field(default_factory=list)
    json_schema: dict[str, Any] = Field(default_factory=dict)


class ReferenceSchemaResponse(BaseModel):
    groups: list[ReferenceGroupSchema] = Field(default_factory=list)
    sections: dict[str, ReferenceSectionSchema] = Field(default_factory=dict)
